### PR TITLE
[Grouping & Mapping] Group Extension

### PR DESCRIPTION
**Pull request by bingjiez**

### CODE REVIEW

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx [hunk 1]
```diff
@@ -58,14 +75,18 @@ const GroupAction = ({
   const [validator, showValidationMessage] = useValidator();
   const [isLoading, setIsLoading] = useState<boolean>(false);
   const [isRendering, setIsRendering] = useState<boolean>(false);
-  const [currentPropertyList, setCurrentPropertyList] = React.useState<PropertyRecord[]>([]);
+  const [currentPropertyList, setCurrentPropertyList] = React.useState<
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:78]: Inconsistent use of `useState`. Everywhere else above you do not use the `React.` prefix.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx [hunk 2]
```diff
@@ -58,14 +75,18 @@ const GroupAction = ({
...
+    PropertyRecord[]
+  >([]);
   const [queryBuilder, setQueryBuilder] = React.useState<QueryBuilder>(
     new QueryBuilder(undefined),
   );
-  const [queryGenerationType, setQueryGenerationType] = React.useState("Selection");
-  const [searchInput, setSearchInput] = React.useState("");
+  const [localQueryGenerationType, setQueryGenerationType] =
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:84]: I would prefer having the state variable as `queryGenerationType`, since usually the state variable and setter are of the pattern `state` and `setState`. I would drop the destructuring of props from parameters of this component in favor of accessing those properties via object, e.g. `props.queryGenerationType`.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx [hunk 3]
```diff
@@ -85,17 +106,24 @@ const GroupAction = ({
...
-          const selection = selectionProvider.getSelection(evt.imodel, evt.level);
-          const query = selection.instanceKeys.size > 0 ? `SELECT ECInstanceId FROM ${selection.instanceKeys.keys().next().value}` : "";
+        if (localQueryGenerationType === "Selection") {
+          const selection = selectionProvider.getSelection(
+            evt.imodel,
+            evt.level,
+          );
+          const query =
+            selection.instanceKeys.size > 0
+              ? `SELECT ECInstanceId FROM ${selection.instanceKeys.keys().next().value
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:116]: Do not split the string into multiple lines. This just makes it more difficult to read. For ternary operators I usually use this formatting:
```
condition
  ? { execute when condition is true }
  : { execute when condition is false }
```

Also, the condition can be on the same line as `const query`.

File: packages/itwin/grouping-mapping-widget/src/widget/utils.ts [hunk 4]
```diff
@@ -2,14 +2,60 @@
...
     timer = window.setTimeout(f, delay, ...args);
   };
 }
+
+/**
+ * Group Extension type definition
+ */
+export interface GroupExtension {
+  /**
+   * unique identifier of the extension
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/utils.ts:27]: All documentation comments should start with a capital letter. All sentences should end with a period.

File: packages/itwin/grouping-mapping-widget/src/widget/utils.ts [hunk 5]
```diff
@@ -2,14 +2,60 @@
...
+   */
+  name: string;
+  /**
+   * display label in the widget
+   */
+  displayLabel: string;
+  /**
+   * UI Component interface
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/utils.ts:35]: This documentation is not enough. To me it is unclear what kind of UI this is and why must it be capitalized? Does all text in that component need to be in capital letters?

File: packages/itwin/grouping-mapping-widget/src/widget/components/context/GroupExtensionContext.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+export const GroupExtensionContext = createContext<IGroupExtension>(
+  createDefaultGroupExtension(),
+);
+
+export const useGroupExtension = () => {
+  const context = useContext(GroupExtensionContext);
+  if (!context) {
+    throw new Error(
+      "useGroupExtension should be used within a GroupExtensionContext provider",
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/context/GroupExtensionContext.ts:37]: Do we really want to make this provider mandatory? I think we should make it optional. If you do not need to add any custom group creation options, why would you create this provider?

File: packages/itwin/grouping-mapping-widget/src/widget/components/extension/ManualExtension.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import "./ManualExtension.scss";
+
+const ManualExtension = ({ updateQuery, isUpdating, resetView }: GroupExtensionProps) => {
+  const [manualInput, setManualInput] = React.useState("");
+
+  return (
+    <div className="gmw-manual-form">
+      <Text>
+        Generate group by user defined ECSQL query. Please select
+        ECInstanceId column in the query.
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/extension/ManualExtension.tsx:18]: `Please select ECInstanceId and ECClassId columns in the query.`. This is a small thing, but it makes the extraction a bit more efficient.

File: packages/itwin/grouping-mapping-widget/src/widget/components/extension/ManualExtension.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      </Text>
+      <LabeledTextarea
+        label="Query"
+        required
+        value={manualInput}
+        onChange={(event) => setManualInput(event.target.value)}
+        disabled={isUpdating}
+        placeholder={`E.g. "Select ECInstanceId From Biscore.Element`}
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/extension/ManualExtension.tsx:26]: Placeholder is missing the ending double quote. Also, change the query to `SELECT ECInstanceId, ECClassId FROM BisCore:PhysicalElement`.

BisCore.Element is a bit too broad and it could select elements that are not displayed in the viewer.

File: common/changes/@itwin/grouping-mapping-widget/gm-extension_2022-09-07-05-13.json [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@itwin/grouping-mapping-widget",
+      "comment": "Rename to customUIProviders and default providers will be replaced if custom ones are provided",
```
Comment (MindaugasButkus) [inline common/changes/@itwin/grouping-mapping-widget/gm-extension_2022-09-07-05-13.json:5]: Why did you add a separate changelog entry for this? I don't think it's a rename if that functionality is being added.

File: packages/itwin/grouping-mapping-widget/README.md [hunk 10]
```diff
@@ -29,3 +29,45 @@ import { GroupingMappingProvider } from "@itwin/grouping-mapping-widget";
...
+`resetView` enables you to reset the viewer state.
+
+To configure your own UI provider, you need to give it:
+
+- a `name` as identifier
+- a `displayLabel` which will show in the 'Add Group' dropdown list
+- a `uiComponent` like above
+- an optional `icon`
+
+You can define your UI provider as following in the GroupingMappingProvider:
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/README.md:60]:
```suggestion
You can define your UI provider as follows in the GroupingMappingProvider:

```

File: packages/itwin/grouping-mapping-widget/src/widget/utils.ts [hunk 11]
```diff
@@ -2,14 +2,60 @@
...
   return (...args: any[]) => {
     timer && window.clearTimeout(timer);
     timer = window.setTimeout(f, delay, ...args);
   };
 }
+
+/**
+ * Custom UI provider type definition
+ */
+export interface CustomUIProvider {
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/utils.ts:25]: It's weird to see that such important interfaces live in the `utils.ts` file. I think they deserve their own file. `CustomUIProvider.ts`?

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/ManualUIProvider.scss [hunk 12]
```diff
@@ -0,0 +1,17 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+@import "~@itwin/itwinui-css/scss/variables";
+
+.gmw-manual-form {
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/provider/ManualUIProvider.scss:7]: Both ManualUIProvider and SearchUIProvider forms share identical styling. Could it reuse the same stylesheet/class or is it a bad practice?

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx [hunk 13]
```diff
@@ -25,34 +35,54 @@ export interface GroupingMappingProps {
    * A custom implementation of MappingClient.
    */
   client?: IMappingsClient;
+  /**
+   * Custom UI providers to add and update groups
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx:39]: Missing period at the end of this sentence.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx [hunk 14]
```diff
@@ -133,43 +164,14 @@ const GroupAction = ({
...
-    ).join(" ")})) OR (${searchQuery.map((token, index) =>
-      `${index === 0 ? "" : needsAndOperator(token, index, searchQuery) ? "AND" : "OR"} ecc.name LIKE '%${isWrappedInQuotes(token) ? token.slice(1, -1) : token}%'`
-    ).join(" ")})`;
-    // Physical Types
-    generatedSearchQuery += ` OR ((${searchQuery.map((token, index) =>
-      `${index === 0 ? "" : needsAndOperator(token, index, searchQuery) ? "AND" : "OR"} pt.codevalue LIKE '%${isWrappedInQuotes(token) ? token.slice(1, -1) : token}%'`
-    ).join(" ")}) OR (${searchQuery.map((token, index) =>
-      `${index === 0 ? "" : needsAndOperator(token, index, searchQuery) ? "AND" : "OR"} pt.userlabel LIKE '%${isWrappedInQuotes(token) ? token.slice(1, -1) : token}%'`
-    ).join(" ")})) `;
+  const updateQuery = useCallback(
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:167]: This callback is redundant. You can pass around the `setQuery` function. It is as stable as the one you create here.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx [hunk 15]
```diff
@@ -218,7 +220,69 @@ const GroupAction = ({
...
+      }
+      default: {
+        if (localQueryGenerationType && localQueryGenerationType.length > 0) {
+          const selectedUIProvider = uiProviders.find(
+            (e) => e.name === localQueryGenerationType,
+          );
+          if (selectedUIProvider) {
+            return React.createElement(selectedUIProvider.uiComponent, {
+              updateQuery,
+              isUpdating: isLoading || isRendering,
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:270]: You already have a local variable called `isUpdating` which evaluates to the same value as this. Reuse it.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx [hunk 16]
```diff
@@ -281,64 +345,53 @@ const GroupAction = ({
         </Fieldset>
         <Fieldset legend='Group By' className='gmw-query-builder-container'>
           <RadioTileGroup
-            className="gmw-radio-group-tile"
+            className='gmw-radio-group-tile'
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:348]: Inconsistent. I see that in most cases double quotes are used, so please stick to double quotes.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx [hunk 17]
```diff
@@ -281,64 +345,53 @@ const GroupAction = ({
...
-              <GroupQueryBuilderContainer />
-            </GroupQueryBuilderContext.Provider> :
-            <div className="gmw-search-form">
-              <Text>Generate a query by keywords. Keywords wrapped in double quotes will be considered a required criteria.</Text>
-              <LabeledTextarea
-                label="Query Keywords"
-                required
-                value={searchInput}
-                onChange={(event) => setSearchInput(event.target.value)}
+            {(uiProviders === undefined || uiProviders.length === 0) && (
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:350]: `uiProviders` will never be undefined, the type does not allow it.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx [hunk 18]
```diff
@@ -281,64 +345,53 @@ const GroupAction = ({
...
+                  name={"groupby"}
+                  onChange={changeGroupByType}
+                  value={"Manual"}
+                  label={"Manual Query"}
+                  disabled={isLoading || isRendering}
+                  checked={localQueryGenerationType === "Manual"}
+                />
+              </>
+            )}
+            {uiProviders?.map((ext) => (
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:381]: `uiProviders` will never be null or undefined, so you can drop the optional chaining operator.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx [hunk 19]
```diff
@@ -281,64 +345,53 @@ const GroupAction = ({
...
-            <div className="gmw-search-form">
-              <Text>Generate a query by keywords. Keywords wrapped in double quotes will be considered a required criteria.</Text>
-              <LabeledTextarea
-                label="Query Keywords"
-                required
-                value={searchInput}
-                onChange={(event) => setSearchInput(event.target.value)}
+            {(uiProviders === undefined || uiProviders.length === 0) && (
+              <>
+                <RadioTile
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:352]: A lot of code duplication happening here. Please reduce.

File: packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.scss [hunk 20]
```diff
@@ -52,4 +52,19 @@
       border: 2px solid black;
     }
   }
+
+  .gmw-menu-item {
+    padding-left: "16px";
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.scss:57]: Try to use less magic numbers. Use a multiple of `$iui-baseline`. Also drop the quotes.

File: packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx [hunk 21]
```diff
@@ -242,12 +250,14 @@ export const Groupings = ({ mapping, goBack }: GroupsTreeProps) => {
     [groups, hiddenGroupsIds, getHiliteIdsFromGroups],
   );
 
-  const addGroup = () => {
+  const addGroup = (type: string) => {
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx:253]: Why does this not use the `useCallback` hook?

File: packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx [hunk 22]
```diff
@@ -390,47 +402,91 @@ export const Groupings = ({ mapping, goBack }: GroupsTreeProps) => {
...
+          <Surface className='gmw-groups-container'>
+            <div className='gmw-toolbar'>
+              <DropdownMenu
+                className='gmw-ui-provider-dropdown'
                 disabled={isLoadingQuery}
-                onClick={addGroup}
+                menuItems={() =>
+                  uiProviders.length > 0
+                    ? uiProviders.map((p) => (
+                      <MenuItem
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx:414]: Please reduce code duplication. You could do some kind of concatenation to render all `MenuItem`s in a single `.map`.

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/ManualUIProvider.tsx [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+  updateQuery,
+  isUpdating,
+  resetView,
+}: CustomUIProviderProps) => {
+  const [manualInput, setManualInput] = React.useState("");
+
+  return (
+    <div className='gmw-manual-form'>
+      <Text>
+        Generate group by user defined ECSQL query. Please select ECInstanceId
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/provider/ManualUIProvider.tsx:21]: `Please select ECInstanceId and ECClassId columns in the query`.

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/SearchUIProvider.tsx [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+  };
+
+  const needsAndOperator = (
+    token: string,
+    index: number,
+    searchQuery: string[],
+  ) =>
+    isWrappedInQuotes(token) ||
+    (index === 1 && isWrappedInQuotes(searchQuery[0]));
+  // Temporary until ECViews become available for use.
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/provider/SearchUIProvider.tsx:28]: Separate functions with an empty line. `needsAndOperator` and `generateSearchQuery` do not have an empty line between them.

File: packages/itwin/grouping-mapping-widget/src/widget/utils.ts [hunk 25]
```diff
@@ -2,14 +2,60 @@
...
+   */
+  icon?: ReactElement;
+}
+
+/**
+ * Custom UI Provider props
+ */
+export interface CustomUIProviderProps {
+  /**
+   * to validate and update current query
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/utils.ts:50]: In documentation start sentences with a capital letter and end with a period.

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/ManualUIProvider.scss (hunk 12 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/ManualUIProvider.scss:7]: I think it's better to separate as it's also an example for user.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx (hunk 14 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:167]: outdated

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx (hunk 15 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:270]: good catch

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx (hunk 16 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:348]: I remember either eslint or prettier use single quote for component attribute; double quote for rest of strings. @Guillar1 Another formatting problem

File: common/changes/@itwin/grouping-mapping-widget/gm-extension_2022-09-07-05-13.json (hunk 9 above)
Comment (MindaugasButkus) [reply common/changes/@itwin/grouping-mapping-widget/gm-extension_2022-09-07-05-13.json:5]: I still don't understand why didn't you simply change the existing changelog and added a new one? In a single release you add something and in the same release you already rename it. The rename does not require a changelog entry, because the end user will see it as a single Add, not Add + Rename.

File: packages/itwin/grouping-mapping-widget/README.md [hunk 26]
```diff
@@ -29,3 +29,45 @@ import { GroupingMappingProvider } from "@itwin/grouping-mapping-widget";
   uiProviders={[new GroupingMappingProvider()]}
 />
 ```
+
+## UI Configuration
+
+We provide interface to define your own group creation method. You can use your own UI component to create/update query and visualize in the viewer accordingly.
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/README.md:35]:
```suggestion
This package provides an interface to define your own group creation method. You can use your own UI component to create or update query and visualize in the viewer accordingly.

```

File: packages/itwin/grouping-mapping-widget/README.md [hunk 27]
```diff
@@ -29,3 +29,45 @@ import { GroupingMappingProvider } from "@itwin/grouping-mapping-widget";
   uiProviders={[new GroupingMappingProvider()]}
 />
 ```
+
+## UI Configuration
+
+We provide interface to define your own group creation method. You can use your own UI component to create/update query and visualize in the viewer accordingly.
+
+Here is an example of group custom UI provider:
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/README.md:37]: Instead of writing this example in README file, you could simply add a link to the `ManualUIProvider.tsx` file, which actually does something. I don't even know if this example would be a valid UI component.

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+  /**
+   * Optional icon, will be shown before display label in widget
+   */
+  icon?: ReactElement;
+}
+
+/**
+ * Custom UI Provider props
+ */
+export interface CustomUIProviderProps {
```
Comment (MindaugasButkus) [inline packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:32]: I think these props are not for the provider, but for the `uiComponent` itself. The name should be changed.

File: packages/itwin/grouping-mapping-widget/src/widget/utils.ts (hunk 4 above)
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/utils.ts]: You did change the first letters to capital letters, but you did not end the documentation sentences with periods.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx (hunk 14 above)
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx:162]: No, it's not outdated.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx (hunk 2 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx]: resolved

**Comment by bingjiez** [comment]
@aruniverse could you review this PR please? thanks

**Comment by aruniverse** [comment]
> @aruniverse could you review this PR please? thanks

I can review after all existing comments have been closed out by @MindaugasButkus

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx (hunk 15 above)
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx]: You missed this in `getRadioTileComponent`.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx (hunk 16 above)
Comment (MindaugasButkus) [reply packages/itwin/grouping-mapping-widget/src/widget/components/GroupAction.tsx]: Do we use prettier?

File: packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx (hunk 21 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/Grouping.tsx:272]: there is no heavy dependency

**Review by MindaugasButkus — APPROVED** [review]
@aruniverse, it's ready for your review.

File: packages/itwin/grouping-mapping-widget/README.md [hunk 29]
```diff
@@ -29,3 +29,37 @@ import { GroupingMappingProvider } from "@itwin/grouping-mapping-widget";
   uiProviders={[new GroupingMappingProvider()]}
 />
 ```
+
+## UI Configuration
+
+This package provides an interface to define your own group creation method. You can use your own UI component to create or update the query.
+
+An example of group custom UI provider, [ManualUIComponent](./src/widget/components/provider/ManualUIComponent.tsx) helps user to manually define ECSQL query for groups.
```
Comment (aruniverse) [inline packages/itwin/grouping-mapping-widget/README.md:37]: might want to make this an absolute path to the file on gh

File: common/changes/@itwin/grouping-mapping-widget/gm-extension_2022-08-17-22-15.json [hunk 30]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@itwin/grouping-mapping-widget",
+      "comment": "Add custom UI provider extensibility and refactor default providers accordingly",
+      "type": "patch"
```
Comment (aruniverse) [inline common/changes/@itwin/grouping-mapping-widget/gm-extension_2022-08-17-22-15.json:6]: minor

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts [hunk 31]
```diff
@@ -0,0 +1,45 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import type { ReactElement } from "react";
+
+/**
+ * Custom UI provider type definition.
+ */
+export interface CustomUIProvider {
```
Comment (aruniverse) [inline packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: do you think tjis will confuse users who are already familiar with ui provider from core? is there a better name for this?

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/ManualUIComponent.tsx [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import React from "react";
+import { Button, LabeledTextarea, Text } from "@itwin/itwinui-react";
+import { LoadingSpinner } from "../utils";
+import "./ManualUIComponent.scss";
+import type { CustomUIComponentProps } from "./CustomUIProvider";
+
+const ManualUIComponent = ({
```
Comment (aruniverse) [inline packages/itwin/grouping-mapping-widget/src/widget/components/provider/ManualUIComponent.tsx:11]: better name?

File: packages/itwin/grouping-mapping-widget/src/widget/components/context/CustomUIProviderContext.ts [hunk 33]
```diff
@@ -0,0 +1,33 @@
...
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import { createContext, useContext } from "react";
+import type { CustomUIProvider } from "../provider/CustomUIProvider";
+
+export const createDefaultCustomUIProvider = (): CustomUIProvider[] => {
+  return [];
+};
+
+export const createCustomUIProvider = (
```
Comment (aruniverse) [inline packages/itwin/grouping-mapping-widget/src/widget/components/context/CustomUIProviderContext.ts:12]: what is the point of this??

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts (hunk 31 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: That's why it's called custom UI provider and it is only used when configuring grouping mapping widget.

File: packages/itwin/grouping-mapping-widget/src/widget/components/context/CustomUIProviderContext.ts (hunk 33 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/context/CustomUIProviderContext.ts:12]: Using context avoids passing in arguments all the way down to grouping component

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts (hunk 31 above)
Comment (aruniverse) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: i dont think adding custom is enough, thats too generic

File: packages/itwin/grouping-mapping-widget/src/widget/components/context/CustomUIProviderContext.ts (hunk 33 above)
Comment (aruniverse) [reply packages/itwin/grouping-mapping-widget/src/widget/components/context/CustomUIProviderContext.ts:12]: the entire createDefaultCustomuiProvider fn is pointless imo, it just returns an empty array.
feel like this file can bec leaned up further

File: packages/itwin/grouping-mapping-widget/README.md (hunk 29 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/README.md:37]: why do we want absolute path? It won't work in cospace, is it?

Comment (aruniverse) [reply packages/itwin/grouping-mapping-widget/README.md:37]: this relative path is only relevant when you have the repo open
when this is ublished to npm or while looking at this when the pkg is isntalled in node_modules, this path wont work

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts (hunk 31 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: hmm any suggestions?

Comment (aruniverse) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: sorry no

Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: How about GMCustomUIProvider?

Comment (aruniverse) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: better, thats on your team to decide tho

Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: We have discussed about this and chose this name after the 'extension' name was rejected.

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/ManualUIComponent.tsx (hunk 32 above)
Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/ManualUIComponent.tsx:11]: ManualUIComponent is for manual query group builder. It follows the same name pattern as other. Detail explanation is in readme

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts (hunk 31 above)
Comment (aruniverse) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: what was this renamed to??

Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: GroupingMappingUIProvider, will push change soon

Comment (aruniverse) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts:10]: not to be confused with `GroupingMappingProvider` which implements `UiItemsProvider` ?

Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts]: Done

**Comment by bingjiez** [comment]
@aruniverse ready to go 🚀

File: packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts (hunk 31 above)
Comment (aruniverse) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts]: I really don't think this rename is enough. I can see this being very confusing if core defines what a UI provider is and you guys piggy back off that name and change it entirely. 

@arnobmallickbsw @Guillar1 @MindaugasButkus can you guys help come up with a less confusing name?

Comment (bingjiez) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts]: GroupingMappingCustomUI? Or GroupingMappingWidgetUI
I prefer the first one.

Comment (aruniverse) [reply packages/itwin/grouping-mapping-widget/src/widget/components/provider/CustomUIProvider.ts]: the first one sounds much better than what we have now, so lets go with that. pleasse replace any usage of Uiprovider thats not pertinent to the UiItemsProvider. so make sure you scrub the changelog file as well.

File: packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx [hunk 34]
```diff
@@ -25,34 +34,49 @@ export interface GroupingMappingProps {
    * A custom implementation of MappingClient.
    */
   client?: IMappingsClient;
+  /**
+   * Custom UI providers to add and update groups.
```
Comment (aruniverse) [inline packages/itwin/grouping-mapping-widget/src/widget/components/GroupingMapping.tsx:38]:
```suggestion
   * Custom UI to add and update groups.
```

**Review by aruniverse — APPROVED** [review]
